File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/q/ping.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.PingInput;
type Output = DistributedCache.PingOutput;

export class PingCache extends BaseHandler<Input, Output> implements DistributedCache.IPingHandler {
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    const start = performance.now();
    try {
      await this.redis.ping();
      const latencyMs = Math.round(performance.now() - start);
      return D2Result.ok({ data: { healthy: true, latencyMs } });
    } catch (err) {
      const latencyMs = Math.round(performance.now() - start);
      const error = err instanceof Error ? err.message : "Unknown error";
      return D2Result.ok({ data: { healthy: false, latencyMs, error } });
    }
  }
}

File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/Q/Ping.cs
// -----------------------------------------------------------------------
// <copyright file="Ping.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.Q;

using System.Diagnostics;
using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;

using H = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.IPingHandler;
using I = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.PingInput;
using O = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead.PingOutput;

/// <summary>
/// Handler for pinging Redis to verify connectivity.
/// </summary>
public class Ping : BaseHandler<H, I, O>, H
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Ping"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Ping(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            var db = r_redis.GetDatabase();
            await db.PingAsync();
            sw.Stop();

            return D2Result<O?>.Ok(
                new O(true, sw.ElapsedMilliseconds, null));
        }
        catch (RedisException ex)
        {
            sw.Stop();
            Context.Logger.LogError(
                ex,
                "Redis ping failed. TraceId: {TraceId}",
                TraceId);

            return D2Result<O?>.Ok(
                new O(false, sw.ElapsedMilliseconds, ex.Message));
        }
    }
}
